lucene/core/src/java/org/apache/lucene/search/FieldComparator.java

```diff
@@ -928,4 +928,9 @@ public int compareTop(int doc) throws IOException {
     @Override
     public void setScorer(Scorable scorer) {}
   }
+
+  public static abstract class IteratorSupplierComparator<T> extends FieldComparator<T> implements LeafFieldComparator {
+    abstract DocIdSetIterator iterator();
+    abstract void updateIterator() throws IOException;
```

Comment:
Indeed it is more straightforward to just update an iterator in `setBottom` function of a comparator.

But I was thinking it is better to have a special function for two reasons:
1)  After updating an iterator, in `TopFieldCollector` we need to change 
`totalHitsRelation = TotalHits.Relation.GREATER_THAN_OR_EQUAL_TO;`

2)  we also need to check `hitsThresholdChecker.isThresholdReached()`, and passing not strictly related object `hitsThresholdChecker` to a comparator's constructor doesn't look nice to me. 
Please let me know if you think otherwise

Reply:
For 1. we could set the totalHitsRelation when we reach the total hits threshold in the TOP_DOCS mode ? 
For 2. I wonder if we could pass the hitsThresholdChecker to the LeafFieldComparator like we do for the scorer ?
This way we can update the iterator internally when a new bottom is set or when `compareBottom` is called ?

Reply:
@msokolov  Thanks for the suggestion, naming is tough, addressed in 95e1bc1.